Ensure e2e tests utilize webhook

Currently we set the webhook failurepolicy in the e2e tests to ignore failures, which allows tests to go through but means we may not be exercising the webhook even when we want to. We should change the failure policy to Fail to ensure that we are actually using the webhook. 

Currently it looks like we start running tests before the operator has had a chance to actually start, so e2e tests fail. I'm not sure what the best way to resolve it would be, perhaps waiting to make sure the operator pods are ready / endpoints are available before running the rest of the tests. See the test output here:
https://devops-ci.elastic.co/job/cloud-on-k8s-pr/2049/console
from this attempted switch:
https://github.com/elastic/cloud-on-k8s/pull/2151

```
11:15:23  + make -C build/ci TARGET=e2e-generate-xml ci
11:15:23  make: Entering directory '/var/lib/jenkins/workspace/cloud-on-k8s-pr/build/ci'
11:15:23  make[1]: Entering directory '/var/lib/jenkins/workspace/cloud-on-k8s-pr/build/ci'
11:15:24  6cb26d63016ef66a4c2cdc0dfc125f6d: Pulling from eck/eck-ci
11:15:24  Digest: sha256:87df5991a3195e61c678c152f3507545d02824af348cf3a52035a825f9c5eaf1
11:15:24  Status: Image is up to date for docker.elastic.co/eck/eck-ci:6cb26d63016ef66a4c2cdc0dfc125f6d
11:15:24  docker.elastic.co/eck/eck-ci:6cb26d63016ef66a4c2cdc0dfc125f6d
11:15:25  [31m✖[0m  test/e2e (719ms)
11:15:25  
11:15:25  === [31mFailed[0m
11:15:25  === [31mFAIL[0m: test/e2e TestSmoke/Creating_an_Elasticsearch_cluster_should_succeed (0.01s)
11:15:25  {"level":"error","@timestamp":"2019-11-21T16:14:25.035Z","message":"stopping early","ver":"0.0.0-00000000","error":"test failure","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128\ngithub.com/elastic/cloud-on-k8s/test/e2e/test.StepList.RunSequential\n\t/go/src/github.com/elastic/cloud-on-k8s/test/e2e/test/step.go:42\ngithub.com/elastic/cloud-on-k8s/test/e2e.TestSmoke\n\t/go/src/github.com/elastic/cloud-on-k8s/test/e2e/smoke_test.go:59\ntesting.tRunner\n\t/usr/local/go/src/testing/testing.go:909"}
11:15:25      --- FAIL: TestSmoke/Creating_an_Elasticsearch_cluster_should_succeed (0.01s)
11:15:25          steps_creation.go:25: 
11:15:25              	Error Trace:	steps_creation.go:25
11:15:25              	Error:      	Received unexpected error:
11:15:25              	            	Internal error occurred: failed calling webhook "elastic-es-validation.k8s.elastic.co": Post https://elastic-webhook-server.e2e-p7l1r-elastic-system.svc:443/validate-elasticsearch-k8s-elastic-co-v1beta1-elasticsearch?timeout=30s: no endpoints available for service "elastic-webhook-server"
11:15:25              	Test:       	TestSmoke/Creating_an_Elasticsearch_cluster_should_succeed
```

--- a/Makefile
+++ b/Makefile
@@ -384,6 +384,7 @@ e2e-local:
 		--auto-port-forwarding \
 		--local \
 		--log-verbosity=$(LOG_VERBOSITY) \
+		--ignore-webhook-failures \
 		--test-timeout=$(TEST_TIMEOUT)
 	@E2E_JSON=$(E2E_JSON) test/e2e/run.sh -run "$(TESTS_MATCH)" -args -testContextPath $(LOCAL_E2E_CTX)
 

--- a/config/e2e/global_operator.yaml
+++ b/config/e2e/global_operator.yaml
@@ -185,6 +185,13 @@ spec:
           requests:
             cpu: 100m
             memory: 20Mi
+        readinessProbe:
+          httpGet:
+            path: /validate-elasticsearch-k8s-elastic-co-v1-elasticsearch
+            port: webhook-server
+            scheme: HTTPS
+          failureThreshold: 3
+          periodSeconds: 5
         ports:
         - containerPort: 9443
           name: webhook-server
@@ -213,7 +220,7 @@ webhooks:
         namespace: {{ .GlobalOperator.Namespace }}
         # this is the path controller-runtime automatically generates
         path: /validate-elasticsearch-k8s-elastic-co-v1beta1-elasticsearch
-    failurePolicy: Ignore
+    failurePolicy: {{ if .IgnoreWebhookFailures }}Ignore{{ else }}Fail{{ end }}
     name: elastic-es-validation.k8s.elastic.co
     rules:
       - apiGroups:
@@ -232,7 +239,7 @@ webhooks:
         namespace: {{ .GlobalOperator.Namespace }}
         # this is the path controller-runtime automatically generates
         path: /validate-elasticsearch-k8s-elastic-co-v1-elasticsearch
-    failurePolicy: Ignore
+    failurePolicy: {{ if .IgnoreWebhookFailures }}Ignore{{ else }}Fail{{ end }}
     name: elastic-es-validation.k8s.elastic.co
     rules:
       - apiGroups:
@@ -255,4 +262,4 @@ spec:
     - port: 443
       targetPort: 9443
   selector:
-    control-plane: elastic-operator
+    control-plane: {{ .GlobalOperator.Name }}

--- a/test/e2e/cmd/run/command.go
+++ b/test/e2e/cmd/run/command.go
@@ -20,23 +20,24 @@ import (
 )
 
 type runFlags struct {
-	managedNamespaces   []string
-	e2eImage            string
-	elasticStackVersion string
-	kubeConfig          string
-	operatorImage       string
-	testContextOutPath  string
-	testLicense         string
-	scratchDirRoot      string
-	testRegex           string
-	testRunName         string
-	commandTimeout      time.Duration
-	autoPortForwarding  bool
-	skipCleanup         bool
-	local               bool
-	logToFile           bool
-	logVerbosity        int
-	testTimeout         time.Duration
+	managedNamespaces     []string
+	e2eImage              string
+	elasticStackVersion   string
+	kubeConfig            string
+	operatorImage         string
+	testContextOutPath    string
+	testLicense           string
+	scratchDirRoot        string
+	testRegex             string
+	testRunName           string
+	commandTimeout        time.Duration
+	autoPortForwarding    bool
+	skipCleanup           bool
+	local                 bool
+	logToFile             bool
+	logVerbosity          int
+	testTimeout           time.Duration
+	ignoreWebhookFailures bool
 }
 
 var log logr.Logger
@@ -78,6 +79,7 @@ func Command() *cobra.Command {
 	cmd.Flags().StringVar(&flags.testRunName, "test-run-name", randomTestRunName(), "Name of this test run")
 	cmd.Flags().DurationVar(&flags.testTimeout, "test-timeout", 5*time.Minute, "Timeout before failing a test")
 	cmd.Flags().BoolVar(&flags.logToFile, "log-to-file", false, "Specifies if should log test output to file. Disabled by default.")
+	cmd.Flags().BoolVar(&flags.ignoreWebhookFailures, "ignore-webhook-failures", false, "Specifies if webhook errors should be ignored. Useful when running test locally. False by default")
 	logutil.BindFlags(cmd.PersistentFlags())
 
 	// enable setting flags via environment variables

--- a/test/e2e/cmd/run/command.go
+++ b/test/e2e/cmd/run/command.go
@@ -35,9 +35,9 @@ type runFlags struct {
 	skipCleanup           bool
 	local                 bool
 	logToFile             bool
+	ignoreWebhookFailures bool
 	logVerbosity          int
 	testTimeout           time.Duration
-	ignoreWebhookFailures bool
 }
 
 var log logr.Logger

--- a/test/e2e/cmd/run/run.go
+++ b/test/e2e/cmd/run/run.go
@@ -141,11 +141,12 @@ func (h *helper) initTestContext() error {
 			},
 			ManagedNamespaces: make([]string, len(h.managedNamespaces)),
 		},
-		OperatorImage: h.operatorImage,
-		TestLicense:   h.testLicense,
-		TestRegex:     h.testRegex,
-		TestRun:       h.testRunName,
-		TestTimeout:   h.testTimeout,
+		OperatorImage:         h.operatorImage,
+		TestLicense:           h.testLicense,
+		TestRegex:             h.testRegex,
+		TestRun:               h.testRunName,
+		TestTimeout:           h.testTimeout,
+		IgnoreWebhookFailures: h.ignoreWebhookFailures,
 	}
 
 	for i, ns := range h.managedNamespaces {

--- a/test/e2e/es/naming_test.go
+++ b/test/e2e/es/naming_test.go
@@ -104,7 +104,7 @@ func testRejectionOfLongName(t *testing.T) {
 					err := k.Client.Create(obj)
 					if err != nil {
 						// validating webhook is active and rejected the request
-						require.Contains(t, err.Error(), `admission webhook "validation.elasticsearch.elastic.co" denied the request`)
+						require.Contains(t, err.Error(), `admission webhook "elastic-es-validation.k8s.elastic.co" denied the request`)
 						return
 					}
 

--- a/test/e2e/test/context.go
+++ b/test/e2e/test/context.go
@@ -61,8 +61,9 @@ func initializeContext() {
 
 func defaultContext() Context {
 	return Context{
-		AutoPortForwarding:  false,
-		ElasticStackVersion: defaultElasticStackVersion,
+		AutoPortForwarding:    false,
+		ElasticStackVersion:   defaultElasticStackVersion,
+		IgnoreWebhookFailures: false,
 		GlobalOperator: ClusterResource{
 			Name:      "elastic-global-operator",
 			Namespace: "elastic-system",
@@ -80,20 +81,21 @@ func defaultContext() Context {
 
 // Context encapsulates data about a specific test run
 type Context struct {
-	GlobalOperator      ClusterResource   `json:"global_operator"`
-	NamespaceOperator   NamespaceOperator `json:"namespace_operator"`
-	E2EImage            string            `json:"e2e_image"`
-	E2ENamespace        string            `json:"e2e_namespace"`
-	E2EServiceAccount   string            `json:"e2e_service_account"`
-	ElasticStackVersion string            `json:"elastic_stack_version"`
-	LogVerbosity        int               `json:"log_verbosity"`
-	OperatorImage       string            `json:"operator_image"`
-	TestLicense         string            `json:"test_license"`
-	TestRegex           string            `json:"test_regex"`
-	TestRun             string            `json:"test_run"`
-	TestTimeout         time.Duration     `json:"test_timeout"`
-	AutoPortForwarding  bool              `json:"auto_port_forwarding"`
-	Local               bool              `json:"local"`
+	GlobalOperator        ClusterResource   `json:"global_operator"`
+	NamespaceOperator     NamespaceOperator `json:"namespace_operator"`
+	E2EImage              string            `json:"e2e_image"`
+	E2ENamespace          string            `json:"e2e_namespace"`
+	E2EServiceAccount     string            `json:"e2e_service_account"`
+	ElasticStackVersion   string            `json:"elastic_stack_version"`
+	LogVerbosity          int               `json:"log_verbosity"`
+	OperatorImage         string            `json:"operator_image"`
+	TestLicense           string            `json:"test_license"`
+	TestRegex             string            `json:"test_regex"`
+	TestRun               string            `json:"test_run"`
+	TestTimeout           time.Duration     `json:"test_timeout"`
+	AutoPortForwarding    bool              `json:"auto_port_forwarding"`
+	Local                 bool              `json:"local"`
+	IgnoreWebhookFailures bool              `json:"ignore_webhook_failures"`
 }
 
 // ManagedNamespace returns the nth managed namespace.

--- a/test/e2e/test/elasticsearch/steps_init.go
+++ b/test/e2e/test/elasticsearch/steps_init.go
@@ -9,12 +9,14 @@ import (
 	"testing"
 
 	esv1 "github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/webhook"
 	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
 	"github.com/elastic/cloud-on-k8s/test/e2e/test"
 	"github.com/stretchr/testify/require"
 	corev1 "k8s.io/api/core/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
 )
 
 // InitTestSteps includes pre-requisite tests (eg. is k8s accessible),
@@ -29,7 +31,6 @@ func (b Builder) InitTestSteps(k *test.K8sClient) test.StepList {
 				require.NoError(t, err)
 			},
 		},
-
 		{
 			Name: "Elasticsearch CRDs should exist",
 			Test: func(t *testing.T) {
@@ -42,7 +43,29 @@ func (b Builder) InitTestSteps(k *test.K8sClient) test.StepList {
 				}
 			},
 		},
-
+		{
+			Name: "Webhook endpoint should not be empty",
+			Test: test.Eventually(func() error {
+				if test.Ctx().IgnoreWebhookFailures {
+					return nil
+				}
+				webhookEndpoints := &corev1.Endpoints{}
+				if err := k.Client.Get(types.NamespacedName{
+					Namespace: test.Ctx().GlobalOperator.Namespace,
+					Name:      webhook.WebhookServiceName,
+				}, webhookEndpoints); err != nil {
+					return err
+				}
+				if len(webhookEndpoints.Subsets) == 0 {
+					return fmt.Errorf(
+						"endpoint %s/%s is empty",
+						webhookEndpoints.Namespace,
+						webhookEndpoints.Name,
+					)
+				}
+				return nil
+			}),
+		},
 		{
 			Name: "Remove Elasticsearch if it already exists",
 			Test: func(t *testing.T) {

